### PR TITLE
feat(adapters): add Hasura adapter

### PR DESCRIPTION
**Pull request by arjunyel**
## ☕️ Reasoning

Adding a Hasura adapter. The Hasura team is willing to maintain

## 🧢 Checklist

- [X] Documentation
- [X] Tests
- [X] Ready to be merged

### CODE REVIEW

**Comment by arjunyel** [comment]
Hi friends, anything you need from me to land this? A lot of the code is auto-generated and tree shaken when you build :)

**Comment by balazsorban44** [comment]
Hi, just wanted to give a note, this is interesting for us, just haven't had the time to look into it yet! :bow:

> "The Hasura team is willing to maintain"

:green_heart:

**Comment by azlekov** [comment]
What is the plan here? Any timeframe when the Hasura adapter will be official?

**Comment by ThangHuuVu** [comment]
hi @arjunyel , I pushed some commits to format the PR. 

- [x] Publish it as ESM-only and under @auth/ namespace. Make sure to have correct package.json file and tsconfig.json file
- [x] Add documentation as JSDoc.

To get this PR merged, could you help with the build errors? Sorry, I am still not very familiar yet with Hasura & GraphQL so I didn't touch the code in `src/`.
 I'm curious about the files that you added in the PR. Could you walk me through all the files? I am guessing that we don't need the `gql` and `hasura` directories, but not sure 🤔. Thanks for the PR & sorry for the delay, let's get this done!

**Comment by arjunyel** [comment]
@ThangHuuVu Fixed build error and cleaned up.

We need the `hasura` directory in order to run the tests against a Hasura instance. `gql` directory has been renamed to `lib` and is the output of GraphQL Codegen, basically it's a generated SDK for accessing Hasura.

Btw I am no longer on the Hasura team so I am not sure of their support plans of this. Thanks for all your help friend!

**Comment by ThangHuuVu** [comment]
> Btw I am no longer on the Hasura team so I am not sure of their support plans of this. Thanks for all your help friend!

🫶 

> We need the `hasura` directory in order to run the tests against a Hasura instance. `gql` directory has been renamed to `lib` and is the output of GraphQL Codegen, basically it's a generated SDK for accessing Hasura.

Thanks for the explanation - if we need to ship the codegen output as part of the library, let's run the script before running `tsc`?

File: packages/adapter-hasura/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+  GetUsersQuery,
+  GetUsersQueryVariables,
+  UpdateSessionMutation,
+  UpdateSessionMutationVariables,
+  UpdateUserMutation,
+  UpdateUserMutationVariables,
+} from "./lib/graphql"
+import { formatDateConversion } from "./utils"
+import type { NonNullify } from "./utils"
+import { log } from "console"
```
Comment (ThangHuuVu) [inline packages/adapter-hasura/src/index.ts:71]: unused import, let's remove

File: packages/adapter-hasura/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+ *   ALTER TABLE ONLY sessions
+ *       ADD CONSTRAINT "sessions_userId_fkey" FOREIGN KEY ("userId") REFERENCES public.users(id) ON UPDATE RESTRICT ON DELETE CASCADE;
+ *
+ *   INSERT INTO provider_type (value) VALUES ('credentials'), ('email'), ('oauth'), ('oidc');
+ *
+ *   ALTER TABLE ONLY accounts
+ *       ADD CONSTRAINT "accounts_type_fkey" FOREIGN KEY ("type") REFERENCES public.provider_type(value) ON UPDATE RESTRICT ON DELETE RESTRICT;
+ *   ```
+ *
+ *1. [Track all the tables and relationships in Hasura](https://hasura.io/docs/latest/schema/postgres/using-existing-database/#step-1-track-tablesviews)
```
Comment (ThangHuuVu) [inline packages/adapter-hasura/src/index.ts:156]: Let's write this as a Info block:
:::info
Tips: [Track all the tables and relationships in Hasura](https://hasura.io/docs/latest/schema/postgres/using-existing-database/#step-1-track-tablesviews)
:::

File: packages/adapter-hasura/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+ *
+ *1. Install the necessary packages
+ *
+ *   ```bash npm2yarn2pnpm
+ *   npm install next-auth @next-auth/hasura-adapter graphql graphql-request
+ *   ```
```
Comment (ThangHuuVu) [inline packages/adapter-hasura/src/index.ts:162]: Let's remove this part, we already have a installation section above

**Review by ThangHuuVu — CHANGES_REQUESTED** [review]

**Comment by arjunyel** [comment]
@ThangHuuVu addressed feedback and moved codegen to a build script

**Review by ThangHuuVu — APPROVED** [review]

**Comment by ThangHuuVu** [comment]
@arjunyel could you help check the build error, I wonder if we really need docker for code gen? 🤔

**Comment by arjunyel** [comment]
@ThangHuuVu sorry for all the trouble! Removed Docker from the build step, replaced it by having the GraphQL Schema as a file in the repo. I need the schema for the codegen so its either run a docker container or have the schema in a GraphQL file. Thank you for all the help!

**Pull request merged**